docs/backend-system/building-plugins-and-modules/08-migrating.md

```diff
@@ -6,4 +6,223 @@ sidebar_label: Migration Guide
+
+```ts
+import {
+  coreServices,
+  createBackendPlugin,
+} from '@backstage/backend-plugin-api';
+import { catalogServiceRef } from '@backstage/plugin-catalog-node';
+import { Router } from 'express';
+import { KubernetesBuilder } from './KubernetesBuilder';
+
+export const kubernetesPlugin = createBackendPlugin({
+  id: 'kubernetes',
+  register(env) {
+    env.registerInit({
+      deps: {
+        logger: coreServices.logger,
+        config: coreServices.config,
+        catalogApi: catalogServiceRef,
+        discovery: coreServices.discovery,
+        // The http router service is used to register the router created by the KubernetesBuilder.
+        http: coreServices.httpRouter,
+      },
+      async init({ config, logger, catalogApi, discovery, http }) {
+        const { router } = await KubernetesBuilder.createBuilder({
+          config,
+          logger,
+          catalogApi,
+          discovery,
+        }).build();
+
+        // We registered the router with the http service.
+        http.use(router);
+      },
+    });
+  },
+});
+```
+
+Done! Users of this plugin are now able to import the `kubernetesPlugin` and register it in their backend using
+
+```ts
+// packages/backend/src/index.ts
+import { kubernetesPlugin } from '@backstage/plugin-kubernetes-backend';
+backend.add(kubernetesPlugin);
+```
+
+There's one thing missing that those sharp eyed readers might have noticed, the `clusterSupplier` option is missing from the original plugin. Let's add it and discuss the alternatives.
+
+One alternative is pass the `ClusterSupplier` in as options to the plugin which is quick and easy but not very flexible and hard to evolve without introducing breaking changes as it changes the public API for the plugin. Having complex types passed in directly to the plugin also clutters the backend setup code and makes it harder to read.
+
+Options are primarily used for simple configuration values that are not complex types. In this case we want to allow users to register their own `ClusterSupplier` implementations to the plugin. This is where the new backend system's [ExtensionPoints](fixme.md) but let's look at doing this with options first.
+
+```ts
+/* omitted imports but they remain the same as above */
+
+export interface KubernetesOptions {
+  clusterSupplier?: KubernetesClustersSupplier;
+}
+
+const kubernetesPlugin = createBackendPlugin((options: KubernetesOptions) => ({
+  id: 'kubernetes',
+  register(env) {
+    env.registerInit({
+      deps: {
+        /* omitted imports but they remain the same as above */
+      },
+      async init({ config, logger, catalogApi, discovery, http }) {
+        const { router } = await KubernetesBuilder.createBuilder({
+          config,
+          logger,
+          catalogApi,
+          discovery,
+        })
+          .setClusterSupplier(options.clusterSupplier)
```

Comment:
OH LOOK IT'S FLUENT

Reply:
WOW MUCH JAVA!

Reply:
Yep it's a good thing fluent interfaces exist so that you don't have to deal with the lack of object literals in Java, which otherwise clearly is the superior pattern.